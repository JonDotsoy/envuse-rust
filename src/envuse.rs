#[cfg(feature = "with-js")]
use js_sys::SyntaxError as JSSyntaxError;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
#[cfg(feature = "with-js")]
use wasm_bindgen::JsValue;

#[derive(Debug)]
pub enum ParsedValue {
    String(String),
    Number(u32),
    Boolean(bool),
    Null,
}

trait Parser {
    fn to_parsed_value(
        name: &String,
        variable_type: &Option<String>,
        options_variable_type: &Option<BTreeMap<String, Option<Expression>>>,
        default_value: &Box<Option<Expression>>,
        nullable: &bool,
        intent_value: &Option<String>,
    ) -> ParsedValue;
}

impl Parser for ParsedValue {
    fn to_parsed_value(
        name: &String,
        variable_type: &Option<String>,
        _options_variable_type: &Option<BTreeMap<String, Option<Expression>>>,
        default_value: &Box<Option<Expression>>,
        nullable: &bool,
        intent_value: &Option<String>,
    ) -> Self {
        let transform_type = variable_type.clone().unwrap_or("String".to_string());

        if nullable == &true && intent_value.is_none() && default_value.is_none() {
            return ParsedValue::Null;
        }

        let proposal_value: String = if let Some(v) = intent_value {
            v.clone()
        } else if let Some(Expression::DefaultValue { value, .. }) = default_value.as_ref() {
            value.clone()
        } else {
            panic!("{}", format!("Cannot found env {}", &name));
        };

        match transform_type.to_lowercase().as_str() {
            "string" => ParsedValue::String(proposal_value),
            "number" => {
                ParsedValue::Number(proposal_value.replace("_", "").parse::<u32>().unwrap())
            }
            "boolean" => match proposal_value.to_lowercase().as_str() {
                "on" | "true" | "1" => ParsedValue::Boolean(true),
                _ => ParsedValue::Boolean(false),
            },
            _ => panic!("Type {} is not valid", transform_type),
        }
    }
}

use crate::{
    parser::{
        ast::{Expression, AST},
        span::Span,
        tokenizer::Tokenizer,
    },
    syntax_error::{DebugOptions, SyntaxError},
};

pub struct Evaluate;

pub struct EvaluateOptions;

pub struct ParseOptions {
    pub environment_values: BTreeMap<String, String>,
}

#[derive(Debug)]
pub struct ProgramError {
    pub message: String,
    pub span: Option<Span>,
    pub source: String,
    pub location: Option<String>,
    pub cause: Option<SyntaxError>,
}

impl ProgramError {
    fn create_prepare_context<T: ToString, F: FnOnce() -> Result<Program, SyntaxError>>(
        location: Option<T>,
        source: T,
        handler: F,
    ) -> Result<Program, ProgramError> {
        match handler() {
            Ok(r) => Ok(r),
            Err(error) => Err(ProgramError {
                message: format!("SyntaxError: {}", error.message),
                source: source.to_string(),
                span: Some(error.span.clone()),
                location: location.map(|t| t.to_string()),
                cause: Some(error),
            }),
        }
    }

    pub fn get_message(self) -> String {
        let mut debug_options = DebugOptions::new();
        debug_options.location = self.location.clone();

        self.cause
            .unwrap()
            .debug_payload_configurable(&self.source, &debug_options)
    }
}

#[cfg(feature = "with-js")]
impl From<ProgramError> for JSSyntaxError {
    fn from(error: ProgramError) -> Self {
        Self::new(&error.get_message().clone())
    }
}

#[cfg(feature = "with-js")]
impl From<ProgramError> for JsValue {
    fn from(error: ProgramError) -> Self {
        Self::from(JSSyntaxError::from(error))
    }
}

/// The program is a structure content the filename and the source
#[derive(Debug, Serialize, Deserialize)]
pub struct Program {
    pub location: Option<String>,
    pub source: String,
    pub ast: Expression,
}

pub trait ToEnvs {
    fn to_envs(self) -> BTreeMap<String, Option<String>>;
}

pub trait ToOptionalString {
    fn to_optional_string(self) -> Option<String>;
}

impl<T: ToString> ToOptionalString for Option<T> {
    fn to_optional_string(self) -> Option<String> {
        self.map(|t| t.to_string())
    }
}

impl ToOptionalString for &str {
    fn to_optional_string(self) -> Option<String> {
        Some(self.to_string())
    }
}

impl ToEnvs for BTreeMap<String, Option<String>> {
    fn to_envs(self) -> BTreeMap<String, Option<String>> {
        self
    }
}

impl<T: ToString, D: ToOptionalString, const Z: usize> ToEnvs for [(T, D); Z] {
    fn to_envs(self) -> BTreeMap<String, Option<String>> {
        BTreeMap::from(self.map(|t| (t.0.to_string(), t.1.to_optional_string())))
    }
}

impl Program {
    pub fn parse<T: ToEnvs>(&self, values: T) -> BTreeMap<String, ParsedValue> {
        let envs_values = values.to_envs();

        let expressions = match &self.ast {
            Expression::Document { elements, .. } => elements,
            _ => panic!("AST Parser error, the ast expressions is not Expression::Document"),
        };

        let mut configs: BTreeMap<String, ParsedValue> = BTreeMap::new();

        for expression in expressions {
            let (key, value) = match expression {
                Expression::Variable {
                    span,
                    comment: _,
                    name,
                    variable_type,
                    options_variable_type,
                    default_value,
                    nullable,
                } => (
                    name,
                    ParsedValue::to_parsed_value(
                        name,
                        variable_type,
                        options_variable_type,
                        default_value,
                        nullable,
                        envs_values.get(name).unwrap_or(&None), // envs_values.get(&variable_type),
                    ),
                ),
                _ => todo!("Expression is not supported"),
            };

            configs.insert(key.to_string(), value);
        }

        configs
    }
}

pub fn create_program<T: ToString>(
    source: T,
    location: Option<T>,
) -> Result<Program, ProgramError> {
    let location_val = location.map(|t| t.to_string());
    ProgramError::create_prepare_context(location_val.clone(), source.to_string(), || {
        Ok(Program {
            location: location_val,
            source: source.to_string(),
            ast: AST::parse(Tokenizer::parse(source.to_string())?)?,
        })
    })
}
