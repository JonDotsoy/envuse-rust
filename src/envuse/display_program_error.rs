use crate::{
    errors::{parser_error::ParseError, program_error::ProgramError},
    syntax_error::SyntaxError,
};

use std::error::Error;

pub fn display_program_error<A, T: ToString>(
    result: Result<A, Box<dyn Error>>,
    source: T,
    location_val: Option<String>,
) -> Result<A, Box<dyn Error>> {
    match result {
        Err(error) if error.is::<SyntaxError>() => {
            let syntax_error = error.downcast_ref::<SyntaxError>().unwrap();
            do yeet ProgramError {
                message: format!("SyntaxError: {}", syntax_error.message),
                span: Some(syntax_error.span.clone()),
                source: source.to_string(),
                location: location_val,
                cause: Some(error),
            }
        }
        Err(error) if error.is::<ParseError>() => {
            let syntax_error = error.downcast_ref::<ParseError>().unwrap();
            do yeet ProgramError {
                message: format!("ParseError: {}", syntax_error.message),
                span: Some(syntax_error.span.clone()),
                source: source.to_string(),
                location: location_val,
                cause: Some(error),
            }
        }
        result_program => result_program,
    }
}
