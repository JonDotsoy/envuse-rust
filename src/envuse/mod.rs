pub mod create_program;
pub mod evaluate;
pub mod evaluate_options;
pub mod parse_options;
pub mod program;
pub mod to_custom_transformers;
pub mod to_envs;
pub mod to_optional_string;
