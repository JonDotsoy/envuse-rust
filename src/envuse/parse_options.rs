use std::collections::BTreeMap;

pub struct ParseOptions {
    pub environment_values: BTreeMap<String, String>,
}
