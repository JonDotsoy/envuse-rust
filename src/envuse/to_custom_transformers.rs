use std::{ptr::NonNull, vec};

pub trait ToCustomTransformers {
    fn to_vec(self) -> Vec<String>;
}

impl<T: ToString> ToCustomTransformers for Option<Vec<T>> {
    fn to_vec(self) -> Vec<String> {
        match self {
            Some(a) => a.iter().map(|s| s.to_string()).collect(),
            None => vec![],
        }
    }
}

impl<T: ToString> ToCustomTransformers for Vec<T> {
    fn to_vec(self) -> Vec<String> {
        self.iter().map(|s| s.to_string()).collect()
    }
}

impl<T: ToString, const Z: usize> ToCustomTransformers for [T; Z] {
    fn to_vec(self) -> Vec<String> {
        self.iter().map(|s| s.to_string()).collect()
    }
}
