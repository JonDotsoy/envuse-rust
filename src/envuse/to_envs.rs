use std::collections::BTreeMap;

use super::to_optional_string::ToOptionalString;

pub trait ToEnvs {
    fn to_envs(self) -> BTreeMap<String, Option<String>>;
}

impl ToEnvs for BTreeMap<String, Option<String>> {
    fn to_envs(self) -> BTreeMap<String, Option<String>> {
        self
    }
}

impl<T: ToString, D: ToOptionalString, const Z: usize> ToEnvs for [(T, D); Z] {
    fn to_envs(self) -> BTreeMap<String, Option<String>> {
        BTreeMap::from(self.map(|t| (t.0.to_string(), t.1.to_optional_string())))
    }
}

impl ToEnvs for Option<BTreeMap<String, Option<String>>> {
    fn to_envs(self) -> BTreeMap<String, Option<String>> {
        BTreeMap::new()
    }
}
