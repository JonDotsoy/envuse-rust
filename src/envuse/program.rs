use crate::errors::program_error::ProgramError;

use super::super::parser::ast::Expression;
use super::super::transformers::kinds::boolean_transform::BooleanTransform;
use super::super::transformers::kinds::number_transform::NumberTransform;
use super::super::transformers::kinds::string_transform::StringTransform;
use super::super::transformers::parser::Parser;
use super::super::transformers::transformer_list::TransformerList;
use super::super::transformers::value_types::ValueType;
use super::to_custom_transformers::ToCustomTransformers;
use super::{super::envuse::to_envs::ToEnvs, display_program_error::display_program_error};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// The program is a structure content the filename and the source
#[derive(Debug, Serialize, Deserialize)]
pub struct Program {
    pub location: Option<String>,
    pub source: String,
    pub ast: Expression,
}

impl Program {
    pub fn parse<T, D>(
        &self,
        values: T,
        custom_transformers: D,
    ) -> Result<BTreeMap<String, ValueType>, Box<dyn std::error::Error>>
    where
        T: ToEnvs,
        D: ToCustomTransformers,
    {
        display_program_error(
            self.parse_unwrap(values, custom_transformers),
            self.source.clone(),
            self.location.clone(),
        )
    }

    fn parse_unwrap<T, D>(
        &self,
        values: T,
        custom_transformers: D,
    ) -> Result<BTreeMap<String, ValueType>, Box<dyn std::error::Error>>
    where
        T: ToEnvs,
        D: ToCustomTransformers,
    {
        let mut transformer_list: TransformerList = Default::default();

        transformer_list.insert("unknown", Box::new(StringTransform));
        transformer_list.insert("str", Box::new(StringTransform));
        transformer_list.insert("string", Box::new(StringTransform));
        transformer_list.insert("number", Box::new(NumberTransform));
        transformer_list.insert("bool", Box::new(BooleanTransform));
        transformer_list.insert("boolean", Box::new(BooleanTransform));

        for key in custom_transformers.to_vec() {
            transformer_list.insert(key, Box::new(StringTransform));
        }

        let envs_values = values.to_envs();

        let document = match self.ast.as_document() {
            Some(document) => document,
            _ => {
                do yeet ProgramError::from((
                    self,
                    "AST Parser error, the ast expressions is not Expression::Document",
                ))
            }
        };

        let expressions = &document.elements;

        let mut configs: BTreeMap<String, ValueType> = BTreeMap::new();

        for expression in expressions {
            let variable = match expression.as_variable() {
                Some(variable) => variable,
                _ => do yeet ProgramError::from((self, "Expression is not supported")),
            };

            let value = Parser::to_parse_variable(&transformer_list, &variable, &envs_values)?;

            configs.insert(variable.name.to_string(), value);
        }

        Ok(configs)
    }
}
