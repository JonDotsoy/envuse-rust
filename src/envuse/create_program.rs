use crate::errors::program_error::ProgramError;

use crate::syntax_error::SyntaxError;

use super::super::parser::ast::AST;
use super::super::parser::tokenizer::Tokenizer;
use super::program::Program;

pub fn create_program<T: ToString>(
    source: T,
    location: Option<T>,
) -> Result<Program, Box<dyn std::error::Error>> {
    let location_val = location.map(|t| t.to_string());

    let result_program: Result<Program, Box<dyn std::error::Error>> = try {
        Program {
            location: location_val.clone(),
            source: source.to_string(),
            ast: AST::parse(Tokenizer::parse(source.to_string())?)?,
        }
    };

    match result_program {
        Err(error) if error.is::<SyntaxError>() => {
            let syntax_error = error.downcast::<SyntaxError>().unwrap();
            do yeet ProgramError {
                message: format!("SyntaxError: {}", syntax_error.message),
                span: Some(syntax_error.span.clone()),
                source: source.to_string(),
                location: location_val,
                cause: Some(*syntax_error),
            }
        }
        result_program => result_program,
    }
}
