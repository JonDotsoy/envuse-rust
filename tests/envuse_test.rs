#[cfg(test)]
mod envuse_test {
    use std::collections::BTreeMap;

    use envuse_parser::envuse::create_program;
    use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot};

    #[test]
    fn should_catch_program_error() {
        let program = create_program(r#"FOO "biz""#, None).err().unwrap();

        assert_debug_snapshot!(program);
        assert_snapshot!(program.get_message());
    }

    #[test]
    fn should_catch_program_error_to_js() {
        let error = create_program("\n\n\tFOO 123", Some("/app/.envuse"))
            .err()
            .unwrap();
        assert_snapshot!(error.get_message());

        let error = create_program("\n\n\tFOO 123", None).err().unwrap();
        assert_snapshot!(error.get_message());
    }

    #[test]
    fn should_create_program() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        assert_debug_snapshot!(program);
    }

    #[test]
    fn should_use_of_parse_must_not_break_with_different_syntax() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        program.parse(BTreeMap::from([
            (String::from("AAA"), None),
            (String::from("FOO"), Some(String::from(""))),
            (String::from("JUM"), None),
            (String::from("TAZ"), None),
        ])).unwrap();

        program.parse([("AAA", ""), ("FOO", ""), ("JUM", ""), ("TAZ", "")]).unwrap();

        program.parse([
            ("AAA", None),
            ("FOO", Some("")),
            ("JUM", None),
            ("TAZ", None),
        ]).unwrap();
    }

    #[test]
    fn should_parse_an_string_value() {
        let program = create_program(r#"FOO="biz""#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "BAR")]));
    }

    #[test]
    fn should_parse_an_number_value() {
        let program = create_program(r#"FOO:Number"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "30_000")]));
    }

    #[test]
    fn should_parse_an_boolean_value() {
        let program = create_program(r#"FOO:Boolean"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("FOO", "true")]));
    }

    #[test]
    fn should_parse_an_null_value() {
        let program = create_program(r#"FOO: String?"#, None).unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")]));
    }

    #[test]
    fn should_read_the_default_values() {
        let program = create_program(r#"FOO: String = "Biz""#, None)
            .map_err(|e| panic!("{}", e.get_message()))
            .unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")]));
    }

    #[test]
    fn should_read_the_default_values_2() {
        let program = create_program(r#"FOO = "Biz""#, None)
            .map_err(|e| panic!("{}", e.get_message()))
            .unwrap();

        assert_debug_snapshot!(program.parse([("BAR", "true")]));
    }

    #[test]
    fn should_full_sample() {
        let source = r#"
            FOO: String = "val"
            BAR: Number = 1223
            BIZ: Boolean = "true"
            BLI: String?
        "#;

        let program = create_program(source, Some(".envuse"))
            .map_err(|e| panic!("{}", e.get_message()))
            .unwrap();

        let envs: [(&str, &str); 0] = [];

        assert_yaml_snapshot!(program.parse(envs).unwrap());
    }

    #[test]
    fn should_catch_error_type_unknown() {
        let source = r###"
            FOO: Unknown
        "###;

        let program = create_program(source, None).unwrap();
        program.parse::<[(&str, &str); 0]>([]).unwrap();

        dbg!(program);
    }
}
