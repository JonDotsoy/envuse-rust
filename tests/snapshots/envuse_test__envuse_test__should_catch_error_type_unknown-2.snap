---
source: tests/envuse_test.rs
expression: error
---
ProgramError {
    message: "ParseError: FOO value cannot be null",
    span: Some(
        Span {
            start: 13,
            end: 25,
        },
    ),
    source: "\n            FOO: unknown\n        ",
    location: None,
    cause: Some(
        ParseError {
            message: "FOO value cannot be null",
            span: Span {
                start: 13,
                end: 25,
            },
        },
    ),
}
